pkg/util/kubernetes/apiserver/resourcetypes.go

```diff
@@ -0,0 +1,165 @@
+		}
+	})
+	return cacheErr
+}
+
+// GetResourceType retrieves the resource type for the given kind and group.
+func GetResourceType(kind, apiVersion string) (string, error) {
+	if resourceCache == nil {
+		return "", fmt.Errorf("resource type cache is not initialized")
+	}
+	return resourceCache.getResourceType(kind, apiVersion)
+}
+
+// getResourceType is the instance method to retrieve a resource type.
+func (r *ResourceTypeCache) getResourceType(kind, apiVersion string) (string, error) {
+	group := getAPIGroup(apiVersion)
+	cacheKey := getCacheKey(kind, group)
+
+	// Check the cache
+	r.lock.RLock()
+	resourceType, found := r.kindGroupToType[cacheKey]
+	r.lock.RUnlock()
+	if found {
+		return resourceType, nil
+	}
+
+	// Query the API server and update the cache
+	resourceType, err := r.discoverResourceType(kind, group)
+	if err != nil {
+		return "", err
+	}
+
+	r.lock.Lock()
+	r.kindGroupToType[cacheKey] = resourceType
+	r.lock.Unlock()
+
+	return resourceType, nil
+}
+
+// discoverResourceType queries the Kubernetes API server to discover the resource type.
+func (r *ResourceTypeCache) discoverResourceType(kind, group string) (string, error) {
+	if r.discoveryClient == nil {
+		return "", fmt.Errorf("discovery client is not initialized")
+	}
+	_, apiResourceLists, err := r.discoveryClient.ServerGroupsAndResources()
+	if err != nil && !discovery.IsGroupDiscoveryFailedError(err) {
+		return "", fmt.Errorf("failed to fetch server resources: %w", err)
+	}
+
+	for _, list := range apiResourceLists {
+		for _, resource := range list.APIResources {
+			if !isValidSubresource(resource.Name) {
+				continue
+			}
+			if resource.Kind == kind && getAPIGroup(list.GroupVersion) == group {
+				return trimSubResource(resource.Name), nil
+			}
+		}
+	}
+	return "", fmt.Errorf("resource type not found for kind %s and group %s", kind, group)
+}
+
+// prepopulateCache pre-fills the cache with all resource types from the discovery client.
+func (r *ResourceTypeCache) prepopulateCache() error {
+	_, apiResourceLists, err := r.discoveryClient.ServerGroupsAndResources()
+	if err != nil && !discovery.IsGroupDiscoveryFailedError(err) {
+		return fmt.Errorf("failed to fetch server resources: %w", err)
+	}
+	if discovery.IsGroupDiscoveryFailedError(err) {
+		log.Warnf("Some groups were skipped during discovery: %v", err)
+	}
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	// Proceed with populating the cache for valid resource groups
+	for _, list := range apiResourceLists {
+		for _, resource := range list.APIResources {
+			if !isValidSubresource(resource.Name) {
+				continue
+			}
+			trimmedResourceType := trimSubResource(resource.Name)
+			cacheKey := getCacheKey(resource.Kind, getAPIGroup(list.GroupVersion))
+			r.kindGroupToType[cacheKey] = trimmedResourceType
+		}
+	}
+
+	return nil
+}
+
+func trimSubResource(resourceType string) string {
+	if strings.Contains(resourceType, "/") {
+		return strings.Split(resourceType, "/")[0]
+	}
+	return resourceType
+}
```

Comment:
I think `strings.Index()` can be used here as well:
```suggestion
func trimSubResource(resourceType string) string {
	if index := strings.Index(resourceType, "/"); index > 0 {
		return resourceType[:index]
	}
	return resourceType
}
```